### DISCUSSION TITLE
Multi surface mesh simplification and locked boundaries

### DISCUSSION THREAD

**Discussion by bp-galopin** [zeux/meshoptimizer#690] [General]
Hi,

Imagine this scenario: a hi poly sphere mesh, both half of the sphere use a different material and the vertices texture coordinates diverge at the frontier.

The way to simplify the model with mesh opts would be to run simplify for each half, using locked vertices or the locked boundaries option flag. This work but has the terrible side effect that it can't simplify the frontier, leading to a lot of wasted triangle to maintain it as you try to achieve lower poly count. 

In a magical world, mesh opts could be fed both triangle list to simplify all together ( also would allow to apply the target error across the whole mesh, in the case the given surfaces are not 50-50 or have various tri density ) and be able to consider an edge collapse on the frontier, applying it to both surfaces at the same time to maintain the continuity when doing so.

We are not in a magical world, so what could be the next best approach that would be feasible given the current state of mesh opts ?

**Comment by zeux** [comment]
So multi material simplification might be supported in a first class manner at some point, but it requires some changes to the core that interfere with other more important improvements. Which is to say we aren’t quite in a magical world and might not be in a while.

However, you can actually do multi material simplification without border locking, with a little extra effort:

1. If the vertices on the boundary of two materials are already duplicated - they will need to be to carry different texture coordinates - the boundary will be automatically classified as a seam and simplified in tandem on both sides. This doesn’t require any code changes.
2. Because the triangles will change during simplification, reassigning materials requires a little work. Probably the easiest solution is to record a per-vertex material id based on the triangles (before running simplify), and then recompute triangle materials after simplification based on the recorded vertex materials. This is pretty simple, linear and will work as long as vertices on the boundary are not shared (see p.1)
3. The only complex part is handling of cases when the boundary vertices are in fact shared. This can happen in general in practice, eg you might be using different shaders/textures but the same texture layout and other data. In that case you would need to duplicate the vertices on the boundary before running simplify; without this the boundary may get distorted and the material assignments won’t work. This is not overwhelmingly difficult but requires identifying vertices with conflicting material assignments, duplicating them (which extends the original vertex buffer and changes the input index buffer), and possibly collapsing them after the simplification if necessary.

If your rendering pipeline already happens to store material ids in vertex data (eg so that a single draw call can be used), then steps 2-3 already happened implicitly and simplify will in fact magically work without changes or extra pre/post processing steps.
